ISSUER_PROTOCOL.md

```diff
@@ -0,0 +1,333 @@
+  ECPoint Wp0;
+  ECPoint Wsp0;
+  ...
+  ECPoint T'n;
+  ECPoint S'n;
+  ECPoint Wpn;
+  ECPoint Wspn;
+  uint16 index;
+} DLEQBatchInput; // Used for Batch proof.
+```
+
+### Issue Function
+
+The _Issue_ function corresponds to the **AT.Sig** stage of the PMBTokens protocol.
+
+Inputs:
+
+*   count (the number of tokens being requested, from the IssueRequest)
+*   nonces (the list of nonces to sign)
+*   secretKey (the secret key that should be used to sign this request, determined by the public metadata)
+*   publicKey (the public key corresponding to the secretKey)
+*   keyID (the ID corresponding to this key)
+*   privateMetadata (the boolean value corresponding to the private metadata value)
+*   toIssue (number of tokens to issue)
+
+Outputs:
+
+*   issued (the number of tokens being issued)
+*   keyID (the key used to sign these tokens)
+*   signed (the list of signed nonces)
+*   cs, us, vs, c0, u0, v0, c1, u1, v1 (the list of scalars corresponding to the zero-knowledge proof)
+
+```
+Issue:
+issued = min(count, toIssue)
+xb,yb = secretKey.x1, secretKey.y1
+if privateMetadata == 0:
+  xb,yb = secretKey.x0, secretKey.y0
+signed = []
+T,S,W,Ws = [], [], [], []
+for i in 0..issued:
+  T' = nonces[i]
+  s ←$ {0, 1}λ
+  S' = Hs(T', s)
+  Wp = xb*T' + yb*S'
+  Wsp = secretKey.xs*T' + secretKey.ys*S'
+  T,S,W,Ws += T', S', Wp, Wsp
+  signed += SignedNonce{s, Wp, Wsp}
+proof = DLEQbatched.P((publicKey,T,S,W,Ws),(secretKey.xs, secretKey.ys, xb, yb))
+return (issued, keyID, signed, proof)
+
+DLEQbatched.P((X,T,S,W,Ws),(xs, ys, xb, yb)):
+e0,...,em = Hc(X,T,S,W,Ws)
+Tt = Sum(ej*Tj) // over j from 0 to m
+St = Sum(ej*Sj) // over j from 0 to m
+Wt = Sum(ej*Wj) // over j from 0 to m
+Wst = Sum(ej*Wsj) // over j from 0 to m
+dleqProof = DLEQ2.P((X,T,S,Ws), (xs,ys)) // cs,us,vs
+dleqorProof = DLEQOR2.P((X,T,S,W), (xb,yb)) // c0,c1,u0,u1,v0,v1
+return dleqProof + dleqorProof
+```
+
+Input Serialization:
+
+```
+Trust Token Issuance Request
+struct {
+  opaque point<1..2^16-1>; // X9.62 Uncompressed point.
+} BlindedNonce;
+
+struct {
+  uint16 count;
+  BlindedNonce nonces[count];
+} IssueRequest;
+```
+
+Output Serialization:
+
+
+```
+Trust Token Issuance Response
+struct {
+  opaque s<Nn>; // big-endian bytestring
+  opaque Wp<1..2^16-1>; // X9.62 Uncompressed Wp point.
+  opaque Wsp<1..2^16-1>; // X9.62 Uncompressed Wsp point.
+} SignedNonce;
+opaque Scalar<Ns>; // big-endian bytestring
+
+struct {
+  Scalar cs;
+  Scalar us;
+  Scalar vs;
+  Scalar c0;
+  Scalar c1;
+  Scalar u0;
+  Scalar u1;
+  Scalar v0;
+  Scalar v1;
+} DLEQProof;
+
+struct {
+  uint16 issued;
+  uint32 key_id = keyID;
+  SignedNonce signed[issued];
+  opaque proof<1..2^16-1>; // Length-prefixed form of DLEQProof.
+} IssueResponse;
+```
+
+### Redeem Function
+
+The _Redeem_ function corresponds to the **AT.VerValid** and **AT.ReadBit** stage of the PMBTokens protocol.
+
+Inputs:
+
+*   token (The trust token to redeem)
+*   client\_data (the client data to include in the SRR)
```

Comment:
Suggestion: add a brief explanation of what this might contain or, alternatively, a link to the definition in the design doc

Reply:
Effectively, this is just "client data". It doesn't have a particular form in the issuer protocol.